src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_ALL;
+
+    UA_ByteString *certificates;
+    size_t certificatesSize = 0;
+    certGroup->getTrustList(certGroup, &trustList);
+
+    if(isTrustedCertificate) {
+        certificates = trustList.trustedCertificates;
+        certificatesSize = trustList.trustedCertificatesSize;
+    } else {
+        certificates = trustList.issuerCertificates;
+        certificatesSize = trustList.issuerCertificatesSize;
+    }
+
+    UA_TrustListDataType list;
+    memset(&list, 0, sizeof(UA_TrustListDataType));
+
+    UA_String thumbpr = UA_STRING_NULL;
+    thumbpr.length = (UA_SHA1_LENGTH * 2);
+    thumbpr.data = (UA_Byte*)malloc(sizeof(UA_Byte)*thumbpr.length);
+    for(size_t i = 0; i < certificatesSize; i++) {
+        UA_CertificateUtils_getThumbprint( &certificates[i], &thumbpr);
+        if(compareThumbprint(&thumbprint, &thumbpr)) {
+            UA_ByteString certificate[1];
+            certificate[0] = certificates[i];
+
+            if(isTrustedCertificate) {
+                list.specifiedLists = UA_TRUSTLISTMASKS_TRUSTEDCERTIFICATES;
+                list.trustedCertificates = certificate;
+                list.trustedCertificatesSize = 1;
+            } else {
+                list.specifiedLists = UA_TRUSTLISTMASKS_ISSUERCERTIFICATES;
+                list.issuerCertificates = certificate;
+                list.issuerCertificatesSize = 1;
+            }
+            break;
+        }
+    }
+
+    if(list.specifiedLists != UA_TRUSTLISTMASKS_NONE) {
+        retval = certGroup->removeFromTrustList(certGroup, &list);
+    } else {
+        UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "The certificate to remove was not found");
+        retval = UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_String_clear(&thumbpr);
+    UA_TrustListDataType_clear(&trustList);
+
+    return retval;
+}
+
+static UA_StatusCode
+openTrustList(UA_Server *server,
+              const UA_NodeId *sessionId, void *sessionHandle,
+              const UA_NodeId *methodId, void *methodContext,
+              const UA_NodeId *objectId, void *objectContext,
+              size_t inputSize, const UA_Variant *input,
+              size_t outputSize, UA_Variant *output) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_BYTE]))/*FileMode*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_Byte fileOpenMode = *(UA_Byte*)input[0].data;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(fileOpenMode & (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING)) {
+        if(!server->transaction)
+            server->transaction = UA_GDSTransaction_new(*sessionId);
+
```

Comment:
Well, that's a deviation from the specification. A call to Open with "Write+EraseExisting" shall only return a BADTRANSACTIONPENDING if a transaction is already pending. But it shall not create a new one. This is the task for the CloseAndUpdate method (see https://reference.opcfoundation.org/GDS/v105/docs/7.8.2.3).
And further, any change during Open and CloseAndUpdate have to be cached somewhere else and are only added to a transaction, if CloseAndUpdate is successfully called. In all other cases the changes done so far are silently discarded without ever being part of a transaction.